---
title: "Mapping simulations"
author: "[gwct](https://gwct.github.io/)"
date: "`r format(Sys.time(), '%m/%d/%Y %H:%M:%S %Z')`"
output:
  #html_document:
  rmdformats::robobook:
    highlight: kate
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
    number_sections: true
    toc_depth: 3
---


```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(ggplot2)
library(cowplot)
library(here)
library(reshape2)
library(dplyr)
source("../lib/design.r")
```

[< Back ](index.html)

# Iterative read mapping - Introduction

Reference genome: mm39

NEAT > BWA > GATK

Chromosome 19 only
chr18 length = 90720763
chr19 length = 61420004

Divergence: 4%
Coverage: 20X

4 rounds of mapping

```{r read, out.width="100%", fig.align = "center", warning=FALSE}
summary_file = here("docs", "data", "mm39-iterative-20X-vcf-summary.csv")
sum_stats_chr = read.csv(summary_file, header=T)

sum_stats = sum_stats_chr %>% group_by(iteration) %>% summarize("golden.variants"=sum(golden.variants), "called.variants"=sum(called.variants), "tp"=sum(tp), "fp"=sum(fp), "fn"=sum(fn), "shared.pos"=sum(shared.pos), "no.info"=sum(no.info))

snp_file = here("docs", "data", "mm39-iterative-20X-vcf-snps.csv")
snp_stats_chr = read.csv(snp_file, header=T, comment.char="#")

mapping_file = here("docs", "data", "mm39-iterative-20X-bam-summary.csv")
bam_stats = read.csv(mapping_file, header=T)
```

# Iteration vs. simulated variants

```{r iter-v-sim, out.width="50%", fig.align = "center", warning=FALSE}

div_sim_p = ggplot(sum_stats, aes(x=iteration, y=golden.variants)) +
  geom_point(size=3, color="#666666") +
  geom_smooth(method="lm", se=F, linetype="dashed", color=corecol(pal="wilke", numcol=1)) +
  #scale_y_continuous(expand=c(0,0)) +
  xlab("Mapping iteration") +
  ylab("# of simulated variants") +
  bartheme()

print(div_sim_p)
```

# Iteration vs. called variants

```{r iter-v-called, out.width="50%", fig.align = "center", warning=FALSE}

div_called_p = ggplot(sum_stats, aes(x=iteration, y=called.variants)) +
  geom_point(size=3, color="#666666") +
  geom_smooth(method="lm", se=F, linetype="dashed", color=corecol(pal="wilke", numcol=1)) +
  geom_hline(yintercept=sum_stats$golden.variants) +
  #scale_y_continuous(expand=c(0,0)) +
  xlab("Mapping iteration") +
  ylab("# of called variants") +
  bartheme()

print(div_called_p)
#cat(paste("# Avg. length: ", mean(ratite_data$length)))
#cat(paste("# Median length: ", median(ratite_data$length)))
```

# Simulated variants vs. called variants

```{r sim-v-called, out.width="50%", fig.align = "center", warning=FALSE, eval=FALSE}

one = data.frame("x"=c(0,1167969), "y"=c(0,1167969))

sim_called_p = ggplot(sum_stats, aes(x=golden.variants, y=called.variants)) +
  geom_smooth(aes(color="Fit"), method="lm", se=F, linetype="dashed") +
  geom_abline(aes(slope=1, intercept=0, color="1:1"), size=1, linetype="dashed", show.legend=F) +
  geom_point(size=3, color="#666666") +
  scale_color_manual(values=c("1:1"="#999999", "Fit"=corecol(pal="wilke", numcol=1))) +
  #scale_y_continuous(expand=c(0,0)) +
  xlab("# of simulated variants") +
  ylab("# of called variants") +
  bartheme()

print(sim_called_p)
```

# SNP error rates by iteration

```{r snp-props, out.width="50%", fig.align = "center", warning=FALSE}

sum_stats$total.snps = sum_stats$tp + sum_stats$shared.pos + sum_stats$fp + sum_stats$fn

sum_stats$tp.prop = sum_stats$tp / sum_stats$total.snps
sum_stats$shared.pos.prop = sum_stats$shared.pos / sum_stats$total.snps
sum_stats$fp.prop = sum_stats$fp / sum_stats$total.snps
sum_stats$fn.prop = sum_stats$fn / sum_stats$total.snps

long_data = melt(sum_stats, id.vars=c("iteration", "golden.variants", "called.variants"))
long_data = subset(long_data, variable != "no.info")
long_data = subset(long_data, variable != "fp" & variable != "fn" & variable != "tp" & variable != "shared.pos")

long_data$text = NA

long_data[long_data$variable=="fp.prop",]$text = sum_stats$fp[sum_stats$iteration==long_data[long_data$variable=="fp.prop",]$iteration]
long_data[long_data$variable=="tp.prop",]$text = sum_stats$tp[sum_stats$iteration==long_data[long_data$variable=="tp.prop",]$iteration]
long_data[long_data$variable=="shared.pos.prop",]$text = NA
long_data[long_data$variable=="fn.prop",]$text = sum_stats$fn[sum_stats$iteration==long_data[long_data$variable=="fn.prop",]$iteration]

long_data$variable = factor(long_data$variable, levels=c("fp.prop", "shared.pos.prop", "tp.prop", "fn.prop"))

cols = c(corecol(pal="wilke", numcol=2, offset=1), "#bda988", corecol(pal="wilke", numcol=1))

overlaps_p = ggplot(long_data, aes(x=iteration, y=value, fill=variable, label=text)) +
  geom_bar(stat="identity") +
  geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  #geom_text(aes(label=text2), size=2.3, color="#f2f2f2", nudge_y=0.05) +
  #geom_text(aes(label=text3), size=2.3, color="#f2f2f2", nudge_y=-0.05) +
  #geom_text(aes(label=text4), size=2.3, color="#f2f2f2", nudge_y=0.82) +
  xlab("Mapping iteration") +
  ylab("Proportion of variants") +
  scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.1)) +
  #scale_x_discrete(limits=unique(long_data$divergence)) +
  scale_x_continuous(breaks=unique(long_data$iteration)) +
  scale_fill_manual(guide=guide_legend(reverse=T), labels=c("False positive", "Shared position", "True positive", "False negative"), values=cols) +
  #scale_fill_discrete(, values=cols) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(overlaps_p)

```

```{r snp-rates, out.width="50%", fig.align = "center", warning=FALSE}

cols = c(corecol(pal="wilke", numcol=1, offset=1), corecol(pal="wilke", numcol=1))

snp_rates_p = ggplot(subset(long_data, variable != "tp.prop" & variable != "shared.pos.prop"), aes(x=iteration, y=value, color=variable)) +
  geom_line(size=1.2) +
  geom_point(size=3) +
  xlab("Mapping iteration") +
  ylab("Proportion of variants") +
  scale_x_continuous(breaks=unique(long_data$iteration)) +
  scale_color_manual(guide=guide_legend(reverse=T), labels=c("False positive", "False negative"), values=cols) +
  bartheme() +
  theme(legend.position="bottom")
print(snp_rates_p)

```

# SNP errors and read depth

```{r fp-fn-depth, out.width="50%", fig.align = "center", warning=FALSE}

snp_stats_chr = subset(snp_stats_chr, !is.na(dp))
snp_stats_chr$dp = as.numeric(snp_stats_chr$dp)

dp_p = ggplot(snp_stats_chr, aes(x=type, y=dp)) +
  geom_boxplot() +
  xlab("SNP classification") +
  ylab("Depth") +
  facet_wrap(~iteration) +
  bartheme()
print(dp_p)


```

# SNP errors and mapping quality

```{r fp-fn-mq, out.width="50%", fig.align = "center", warning=FALSE}

snp_stats_chr$mq = as.numeric(snp_stats_chr$mq)

mq_p = ggplot(snp_stats_chr, aes(x=type, y=mq)) +
  geom_boxplot() +
  xlab("SNP classification") +
  ylab("Mapping Quality") +
  facet_wrap(~iteration) +
  bartheme()
print(mq_p)


```

# Read mapping errors

```{r read-props, out.width="50%", fig.align = "center", warning=FALSE}

bam_stats$total.reads = bam_stats$exact.map + bam_stats$close.map + bam_stats$mismapped + bam_stats$diff.chr + bam_stats$unmapped

bam_stats$exact.map.prop = bam_stats$exact.map / bam_stats$total.reads
bam_stats$close.map.prop = bam_stats$close.map / bam_stats$total.reads
bam_stats$mismapped.prop = bam_stats$mismapped / bam_stats$total.reads
bam_stats$diff.chr.prop = bam_stats$diff.chr / bam_stats$total.reads
bam_stats$unmapped.prop = bam_stats$unmapped / bam_stats$total.reads

long_data = melt(bam_stats, id.vars=c("iteration", "exact.map", "close.map", "mismapped", "diff.chr", "unmapped"))
long_data = subset(long_data, variable != "coverage")
long_data = subset(long_data, variable != "divergence")
long_data = subset(long_data, variable != "exact.map" & variable != "close.map" & variable != "mismapped" & variable != "diff.chr" & variable != "unmapped")

long_data$text = NA

long_data[long_data$variable=="exact.map.prop",]$text = bam_stats$exact.map[bam_stats$iteration==long_data[long_data$variable=="exact.map.prop",]$iteration]
long_data[long_data$variable=="close.map.prop",]$text = bam_stats$close.map[bam_stats$iteration==long_data[long_data$variable=="close.map.prop",]$iteration]
long_data[long_data$variable=="mismapped.prop",]$text = bam_stats$mismapped[bam_stats$iteration==long_data[long_data$variable=="mismapped.prop",]$iteration]
long_data[long_data$variable=="diff.chr.prop",]$text = bam_stats$diff.chr[bam_stats$iteration==long_data[long_data$variable=="diff.chr.prop",]$iteration]
long_data[long_data$variable=="unmapped.prop",]$text = bam_stats$unmapped[bam_stats$iteration==long_data[long_data$variable=="unmapped.prop",]$iteration]

long_data$variable = factor(long_data$variable, levels=c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop"))

cols = corecol(pal="wilke", numcol=5, offset=1)

overlaps_p = ggplot(long_data, aes(x=iteration, y=value, fill=variable, label=text)) +
  geom_bar(stat="identity") +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  #geom_text(aes(label=text2), size=2.3, color="#f2f2f2", nudge_y=0.05) +
  #geom_text(aes(label=text3), size=2.3, color="#f2f2f2", nudge_y=-0.05) +
  #geom_text(aes(label=text4), size=2.3, color="#f2f2f2", nudge_y=0.82) +
  xlab("Mapping iteration") +
  ylab("Proportion of reads") +
  scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.1)) +
  #scale_x_discrete(limits=unique(long_data$divergence)) +
  scale_x_continuous(breaks=unique(long_data$iteration)) +
  scale_fill_manual(guide=guide_legend(reverse=T), labels=c("Exact map", "Close map", "Mismapped", "Different chrome", "Unmapped"), values=cols) +
  #scale_fill_discrete(, values=cols) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(overlaps_p)

```

```{r read-rates, out.width="50%", fig.align = "center", warning=FALSE}

cols = corecol(pal="wilke", numcol=4, offset=2)

read_rates_p = ggplot(subset(long_data, variable != "exact.map.prop"), aes(x=iteration, y=value, color=variable)) +
  geom_line(size=1.2) +
  geom_point(size=3) +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  #geom_text(aes(label=text2), size=2.3, color="#f2f2f2", nudge_y=0.05) +
  #geom_text(aes(label=text3), size=2.3, color="#f2f2f2", nudge_y=-0.05) +
  #geom_text(aes(label=text4), size=2.3, color="#f2f2f2", nudge_y=0.82) +
  xlab("Mapping iteration") +
  ylab("Proportion of reads") +
  #scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.1)) +
  #scale_x_discrete(limits=unique(long_data$divergence)) +
  scale_x_continuous(breaks=unique(long_data$iteration)) +
  scale_color_manual(guide=guide_legend(reverse=T), labels=c("Close map", "Mismapped", "Different chrome", "Unmapped"), values=cols) +
  #scale_fill_discrete(, values=cols) +
  bartheme() +
  theme(legend.position="bottom")
print(read_rates_p)

```

[< Back ](index.html)

















