---
title: "Mapping simulations"
author: "[gwct](https://gwct.github.io/)"
date: "`r format(Sys.time(), '%m/%d/%Y %H:%M:%S %Z')`"
output:
  #html_document:
  rmdformats::robobook:
    highlight: kate
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
    number_sections: true
    toc_depth: 3
---


```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(tidyverse)
library(here)
library(cowplot)
library(kableExtra)
library(viridis)
source("../lib/design.r")
```

[< Back ](index.html)

# Introduction

Reference genome: mm39

NEAT (divergence) > bcftools consensus > NEAT (heterozygosity) > BWA > GATK > [3 iterations > minimap > paftools]

- Chromosomes 18
- chr18 length = 90720763

- Divergence: 0-10%
- Heterozygosity: 0.05%
- Coverage: 30X for both NEAT runs

```{r read, out.width="100%", fig.align = "center", warning=FALSE}

vcf_comp_file = here("data", "mm39-30X-0.005h-3i-vcf-comparison.tsv.gz")
vcf_comp = read_tsv(vcf_comp_file, col_names=c("coverage", "divergence", "heterozygosity", "iteration", "chr", "pos", "ref", "golden.div", "prev.iter", "iter.gt", "iter.allele.1", "iter.allele.2", "minimap.gt", "minimap.allele.1", "minimap.allele.2"))
# Load the SNP data (~15gb, takes time)

####################

mapping_file = here("data", "mm39-30X-0.005h-3i-bam-summary.csv")
bam_stats = read.csv(mapping_file, header=T)
# Read the bam compare file

####################

len_18 = 90720763
len_19 = 61420004
total_len = len_18# + len_19
## TODO: Use .fai file to do this automatically in the full sims

####################

```

# Sanity check

Simulated divergence vs. % of sites simulated with SNPs for iteration 1, just to make sure the simulations are working

```{r sim-check, out.width="50%", fig.align = "center", warning=FALSE}

sim_vars_div_iter1 = vcf_comp %>% filter(iteration == 1 & ref != golden.div) %>%
  group_by(divergence) %>%
  summarize(n=n()) %>%
  mutate(perc.var.sites=n/total_len)

div_sim_p = ggplot(sim_vars_div_iter1, aes(x=divergence, y=perc.var.sites)) +
    geom_point(size=4, color="#666666") +
    geom_abline(slope=1, intercept=0, size=1, linetype="dashed", color=corecol(numcol=1, pal="wilke")) +
    scale_y_continuous(limits=c(0,0.10), breaks=seq(0,0.10,0.02)) +
    xlab("Simulated divergence") +
    ylab("% of sites simulated with SNPs") +
    bartheme()
print(div_sim_p)

## Note that with increasing specified divergence, actual divergence is under-simulated

```

# 1 iteration of mapping

## Read mapping errors

"Missing" reads likely those filtered by BWA (option `-T` which defaults to 30: https://bio-bwa.sourceforge.net/bwa.shtml)

```{r read-props, out.width="50%", fig.align = "center", warning=FALSE}

read_classes = c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop", "missing.prop")
read_class_names = c("Exact map", "Close map", "Mismapped", "Different chrome", "Unmapped", "Missing")
# The various classes of reads

bam_stats_iter1 = bam_stats %>% filter(iteration == 1) %>% mutate(total.reads = exact.map + close.map + mismapped + diff.chr + unmapped + missing.in.query) %>%
  mutate(exact.map.prop = exact.map / total.reads, close.map.prop = close.map / total.reads, mismapped.prop = mismapped / total.reads, 
         diff.chr.prop = diff.chr / total.reads, unmapped.prop = unmapped / total.reads, missing.prop = missing.in.query / total.reads)
# Calculate the percent of each type of read from the first iteration of mappping

long_bam_data = bam_stats_iter1 %>% 
  select(iteration, divergence, exact.map.prop, close.map.prop, mismapped.prop, diff.chr.prop, unmapped.prop, missing.prop) %>% 
  pivot_longer(cols=c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop", "missing.prop"))
# Convert to long format

long_bam_data$name = factor(long_bam_data$name, levels=read_classes)
# Factorize the read classes

cols = corecol(pal="wilke", numcol=6, offset=1)
# Get some colors

reads_p = ggplot(long_bam_data, aes(x=divergence, y=value, fill=name)) +
  geom_bar(stat="identity", position=position_fill(reverse = TRUE)) +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  xlab("Simulated ivergence") +
  ylab("Proportion of reads") +
  scale_x_continuous(breaks=seq(0, 0.1, by=0.02)) +
  scale_y_continuous(expand=c(0,0)) +
  scale_fill_manual(labels=read_class_names, values=cols) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(reads_p)

####################

cols = corecol(pal="wilke", numcol=5, offset=2)
names(cols) = read_classes[-1]

read_rates_p = ggplot(subset(long_bam_data, name != "exact.map.prop"), aes(x=divergence, y=value, color=name)) +
  geom_line(size=1.2) +
  geom_point(size=3) +
  xlab("Simulated divergence") +
  ylab("Proportion of reads") +
  scale_x_continuous(breaks=unique(long_bam_data$divergence)) +
  scale_color_manual(labels=read_class_names[-1], values=cols) +
  bartheme() +
  theme(legend.position="bottom")
print(read_rates_p)

```

## Heterozygous sites

```{r hets, out.width="50%", fig.align = "center", warning=FALSE}

cols = c(corecol(numcol=1), "#333333")
names(cols) = c("Called", "Simulated")
# Get some colors for the plot

het_sites = vcf_comp %>% filter(iteration == 1 & iter.gt == "het") %>% 
  group_by(divergence) %>% 
  summarize(n=n()) %>% 
  mutate(perc.het.sites=n/total_len)
# Get sites that were called heterozygous

het_p = ggplot(het_sites, aes(x=divergence, y=perc.het.sites)) +
  #geom_smooth(aes(color="Fit"), method="lm", se=F, linetype="dashed") +
  geom_hline(aes(color="Simulated"), yintercept=0.005, size=1, linetype="dashed", show.legend=F) +
  geom_line(aes(color="Called"), size=1) +
  geom_point(aes(color="Called"), size=3) +
  scale_color_manual(values=cols) +
  scale_y_continuous(limits=c(0,0.0055)) +
  xlab("Simulated divergence") +
  ylab("% of sites\ncalled heterozygous") +
  bartheme() +
  theme(legend.position="none")
print(het_p)

```

## Substitutions

Classified as follows:

- True positive (TP): Simulated variant in original golden VCF and homozygous alternate in called VCF with alternate allele matching the simulated allele
- False positive (FP): Not present in original golden VCF and homozygous alternate in called VCF
- False negative (FN): Simulated variant in original golden VCF and homozygous reference in called VCF with alternate allele matching the reference allele

```{r snp-props, out.width="50%", fig.align = "center", warning=FALSE}

tp = vcf_comp %>% filter(iteration == 1 & golden.div != ref & iter.gt == "hom.alt" & iter.allele.1 == golden.div) %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="tp")
fp = vcf_comp %>% filter(iteration == 1 & golden.div == ref & iter.gt == "hom.alt")  %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="fp")
fn = vcf_comp %>% filter(iteration == 1 & golden.div != ref & iter.gt == "hom.prev" & iter.allele.1 == ref) %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="fn")
# Get sites of different classes from the iteration VCF (GATK)

snp_classes = rbind(tp, fp, fn)
# Combine the various snp classes

snp_classes = snp_classes %>% group_by(divergence, iteration) %>% 
  mutate(total.sites = sum(n)) %>% 
  mutate(prop = n / total.sites)
# Calculate the proportion of each type of site

snp_classes$class = factor(snp_classes$class, levels=c("fn", "tp", "fp"))
cols = c("fp"=corecol(pal="wilke", numcol=1, offset=1), "tp"="#bda988", "fn"=corecol(pal="wilke", numcol=1))
# Factorize the classes and get some colors

snps_p = ggplot(snp_classes, aes(x=divergence, y=prop, fill=class)) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE)) +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  xlab("Simulated divergence") +
  ylab("Proportion of variants") +
  scale_x_continuous(breaks=seq(0, 0.1, by=0.02)) +
  scale_y_continuous(expand=c(0,0)) +
  scale_fill_manual(labels=c("False negative", "True positive", "False positive"), values=cols) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(snps_p)

####################

cols = c(corecol(pal="wilke", numcol=1), corecol(pal="wilke", numcol=1, offset=1))
# Get some colors for this plot

snp_rates_p = ggplot(subset(snp_classes, class != "tp" & divergence > 0), aes(x=divergence, y=prop, color=class)) +
  geom_line(size=1.2) +
  geom_point(size=3) +
  xlab("Simulated divergence") +
  ylab("Proportion of variants") +
  scale_x_continuous(breaks=unique(snp_classes$divergence)) +
  scale_color_manual(labels=c("False negative", "False positive"), values=cols) +
  bartheme() +
  theme(legend.position="bottom")
print(snp_rates_p)

```

# 3 iterations of mapping

NEAT (divergence) > bcftools consensus > NEAT (heterozygosity) > [BWA > GATK > bcftools consensus] x 3

## Read mapping errors

"Missing" reads likely those filtered by BWA (option `-T` which defaults to 30: https://bio-bwa.sourceforge.net/bwa.shtml)

```{r read-props-iter, out.width="50%", fig.align = "center", warning=FALSE}

read_classes = c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop", "missing.prop")
read_class_names = c("Exact map", "Close map", "Mismapped", "Different chrome", "Unmapped", "Missing")
# The various classes of reads

bam_stats = bam_stats %>% mutate(total.reads = exact.map + close.map + mismapped + diff.chr + unmapped + missing.in.query) %>%
  mutate(exact.map.prop = exact.map / total.reads, close.map.prop = close.map / total.reads, mismapped.prop = mismapped / total.reads, 
         diff.chr.prop = diff.chr / total.reads, unmapped.prop = unmapped / total.reads, missing.prop = missing.in.query / total.reads)
# Calculate the proportion of each read type

long_bam_data = bam_stats %>% 
  select(iteration, divergence, exact.map.prop, close.map.prop, mismapped.prop, diff.chr.prop, unmapped.prop, missing.prop) %>% 
  pivot_longer(cols=c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop", "missing.prop"))
# Convert to long format

long_bam_data$name = factor(long_bam_data$name, levels=read_classes)
# Factorize the read classes

cols = corecol(pal="wilke", numcol=6, offset=1)
# Get some colors

overlaps_p = ggplot(long_bam_data, aes(x=iteration, y=value, fill=name)) +
  geom_bar(stat="identity") +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  xlab("Mapping iteration") +
  ylab("Proportion of reads") +
  ggtitle("By divergence") +
  scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.2)) +
  scale_fill_manual(labels=read_class_names, values=cols) +
  facet_wrap(~divergence) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(overlaps_p)

####################

cols = corecol(pal="wilke", numcol=5, offset=2)
names(cols) = read_classes[-1]
# Get some colors for the plot

read_rates_p = ggplot(subset(long_bam_data, name != "exact.map.prop"), aes(x=iteration, y=value, color=name)) +
  geom_line(size=1.2) +
  geom_point(size=3) +
  xlab("Divergence") +
  ylab("Proportion of reads") +
  scale_x_continuous(breaks=unique(long_bam_data$divergence)) +
  scale_color_manual(labels=read_class_names[-1], values=cols) +
  facet_wrap(~divergence) +
  bartheme() +
  theme(legend.position="bottom")
print(read_rates_p)

```

## Substitutions called per iteration

```{r iter-v-called, out.width="50%", fig.align = "center", warning=FALSE}

# called_vars = vcf_comp %>% filter(iter.gt == "hom.alt")
# called_vars_iter = called_vars %>% group_by(divergence, iteration) %>% summarize(n=n()) %>% mutate(perc.called.sites=n/total_len)
# called_vars_iter$iteration = as.character(called_vars_iter$iteration)
# div_called_p = ggplot(called_vars_iter, aes(x=divergence, y=perc.called.sites, color=iteration)) +
#   geom_point(size=3) +
#   xlab("Simulated divergence") +
#   ylab("% of sites with homozygous alt variant called") +
#   bartheme()
# print(div_called_p)
# This is the same plot but compared to the previous iteration consensus

minimap_vars_iter = vcf_comp %>% filter(minimap.gt == "hom.alt" & minimap.allele.1 == golden.div) %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n()) %>% 
  mutate(perc.called.sites=n/total_len)
minimap_vars_iter$iteration = as.character(minimap_vars_iter$iteration)

cols = corecol(numcol=3, pal="wilke")

div_called_p = ggplot(minimap_vars_iter, aes(x=divergence, y=perc.called.sites, color=iteration)) +
  geom_point(size=4, alpha=0.5) +
  geom_abline(slope=1, intercept=0, size=1, linetype="dashed", color="#999999") +
  geom_line(data=sim_vars_div_iter1, aes(x=divergence, y=perc.var.sites), color="#000000") +
  geom_point(data=sim_vars_div_iter1, aes(x=divergence, y=perc.var.sites), size=4, color="#000000") +
  scale_x_continuous(limits=c(0,0.1), breaks=seq(0, 0.1, by=0.02)) +
  scale_y_continuous(limits=c(0,0.1), breaks=seq(0, 0.1, by=0.02)) +
  xlab("Simulated divergence") +
  ylab("% of sites with homozygous alt variant correctly called") +
  scale_color_manual(name="Iteration", values=cols) +
  bartheme() +
  theme(legend.title = element_text(size=12))

print(div_called_p)

####################

minimap_vars_iter$ratio = NA
for(div in levels(as.factor(sim_vars_div_iter1$divergence))){
  minimap_vars_iter[minimap_vars_iter$divergence==div,]$ratio = minimap_vars_iter[minimap_vars_iter$divergence==div,]$perc.called.sites / sim_vars_div_iter1[sim_vars_div_iter1$divergence==div,]$perc.var.sites
}

ratio_called_p = ggplot(minimap_vars_iter, aes(x=divergence, y=ratio, color=iteration)) +
  geom_point(size=4, alpha=0.5) +
  geom_hline(yintercept=1, size=1, linetype="dashed", color="#999999") + 
  scale_x_continuous(limits=c(0.02,0.1), breaks=seq(0, 0.1, by=0.02)) +
  scale_y_continuous(limits=c(0,1)) +
  xlab("Simulated divergence") +
  ylab("Ratio of correctly called variants to simulated variants") +
  scale_color_manual(name="Iteration", values=cols) +
  bartheme() +
  theme(legend.title = element_text(size=12))

print(ratio_called_p)

```

```{r sim-v-called, out.width="50%", fig.align = "center", warning=FALSE, eval=FALSE}

golden_mm_comp = vcf_comp %>% filter(golden.div != ref & minimap.gt == "hom.alt" & minimap.allele.1 == golden.div)
golden_mm_comp_iter = golden_mm_comp %>% group_by(divergence, iteration) %>% summarize(n=n()) %>% mutate(perc.called.sites=n/total_len)
golden_mm_comp_iter$iteration = as.character(golden_mm_comp_iter$iteration)

#one = data.frame("x"=c(0,1167969), "y"=c(0,1167969))

sim_called_p = ggplot(golden_mm_comp_iter, aes(x=divergence, y=perc.called.sites, color=iteration)) +
  geom_point(size=4, alpha=0.5) +
  geom_abline(slope=1, intercept=0, size=1.5, linetype="dashed", color="#999999") +
  scale_x_continuous(limits=c(0.02,0.1), breaks=seq(0, 0.1, by=0.02)) +
  scale_y_continuous(limits=c(0.01,0.1), breaks=seq(0, 0.1, by=0.02)) +
  xlab("Simulated divergence") +
  ylab("% of sites with correct homozygous alt variant called") +
  scale_color_manual(name="Iteration", values=cols) +
  bartheme() +
  theme(legend.title = element_text(size=12))

print(sim_called_p)
```

## Substitution error rates

Substitutions (homozygous alternate alleles) were classified as follows:

- True positive (TP): Variant in original golden VCF and homozygous alternate in called VCF
- False positive (FP): Invariant in original golden VCF and homozygous alternate in called VCF
- False negative (FN): Variant in original golden VCF and homozygous reference in called VCF
- True negative (TN): Invariant in original golden VCF and homozygous reference in called VCF (not shown)

```{r snp-props-iter, out.width="50%", fig.align = "center", warning=FALSE}

tp = vcf_comp %>% filter(golden.div != ref & minimap.gt == "hom.alt" & minimap.allele.1 == golden.div) %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="tp")
fp = vcf_comp %>% filter(golden.div == ref & minimap.gt == "hom.alt") %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="fp")
fn = vcf_comp %>% filter(golden.div != ref & minimap.gt == "hom.ref" & minimap.allele.1 == ref) %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="fn")

snp_classes = rbind(tp, fp, fn)
snp_classes = snp_classes %>% group_by(divergence, iteration) %>% mutate(total.sites = sum(n)) %>% mutate(prop = n / total.sites)
snp_classes$class = factor(snp_classes$class, levels=c("fn", "tp", "fp"))

cols = c("fp"=corecol(pal="wilke", numcol=1, offset=1), "tp"="#bda988", "fn"=corecol(pal="wilke", numcol=1))

overlaps_p = ggplot(snp_classes, aes(x=iteration, y=prop, fill=class)) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE)) +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  xlab("Mapping iteration") +
  ylab("Proportion of variants") +
  ggtitle("By divergence") +
  scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.2)) +
  scale_fill_manual(labels=c("False negative", "True positive", "False positive"), values=cols) +
  facet_wrap(~divergence) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(overlaps_p)

```

```{r snp-rates-iter, out.width="50%", fig.align = "center", warning=FALSE}

cols = c(corecol(pal="wilke", numcol=1), corecol(pal="wilke", numcol=1, offset=1))

snp_rates_p = snp_classes %>%
  filter(class != "tp", divergence > 0) %>%
  ggplot(aes(x=iteration, y=prop, color=class)) +
    geom_line(size=1.2) +
    geom_point(size=3) +
    xlab("Mapping iteration") +
    ylab("Proportion of variants") +
    #scale_x_continuous(breaks=unique(long_data$iteration)) +
    scale_color_manual(labels=c("False negative", "False positive"), values=cols) +
    facet_wrap(~divergence) +
    bartheme() +
    theme(legend.position="bottom")
print(snp_rates_p)

```

## Estimated divergence across the genome

I think this is wrong because bedtools is counting heterozyous and/or filtered sites ... ?

```{r iterative-div, out.width="50%", fig.align = "center", warning=FALSE}

library(zoo)
# For the rolling mean function

window_file = here("data", "window-snps-30X-summary.tab.gz")
  
windows = read_tsv(window_file, col_names=c("chr", "start", "end", "snps", "coverage", "divergence", "heterozygosity", "iteration"))
windows = windows %>% filter(chr %in% c("18")) %>% mutate(length=end-start) %>% mutate(est.div=snps/length)
#windows$iteration = as.factor(windows$iteration)

#x = subset(windows, divergence == 0.06 & iteration == 1)
#sum(x$snps)
# 7527312
#nrow(vcf_comp %>% filter(divergence == 0.06 & iteration == 1 & minimap.gt == "hom.alt"))
# 4309775
## Why so many more SNPs from bedtools? Counting het sites??

p =  windows %>% ggplot(aes(x=start, y=rollmean(est.div, 100, na.pad = TRUE, align = "right"), color=iteration)) +
  geom_line() +
  facet_wrap(~divergence) +
  scale_color_manual(values=corecol(pal="wilke", numcol=4)) +
  bartheme()

print(p)


```

# 1 iteration of mapping + mapping to a genome graph

NEAT (divergence) > bcftools consensus > NEAT (heterozygosity) > BWA > GATK > vg autoindex > vg giraffe

## Read mapping errors

Since the graph pipeline resulted in no "Missing" reads, I removed the "Missing" reads from all BWA mappings for more accurate comparisons...

```{r read-props-graph, out.width="50%", fig.align = "center", warning=FALSE}

graph_file = here("data", "mm39-30X-0.005h-bam-graph-summary.csv")
graph_stats = read.csv(graph_file, header=T)
# Read the graph mapping stats

####################

read_classes = c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop")
read_class_names = c("Exact map", "Close map", "Mismapped", "Different chrome", "Unmapped")
# The various classes of reads

bam_stats = bam_stats %>% mutate(total.reads = exact.map + close.map + mismapped + diff.chr + unmapped) %>%
  mutate(exact.map.prop = exact.map / total.reads, close.map.prop = close.map / total.reads, mismapped.prop = mismapped / total.reads, 
         diff.chr.prop = diff.chr / total.reads, unmapped.prop = unmapped / total.reads)
# Calculate the proportion of each read type for the BWA runs

long_bam_data = bam_stats %>% 
  select(iteration, divergence, exact.map.prop, close.map.prop, mismapped.prop, diff.chr.prop, unmapped.prop) %>% 
  pivot_longer(cols=c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop"))
# Convert to long format

####################

graph_stats$iteration = "1+graph"
# Add a label to the graph stats

graph_stats = graph_stats %>% mutate(total.reads = exact.map + close.map + mismapped + diff.chr + unmapped) %>%
  mutate(exact.map.prop = exact.map / total.reads, close.map.prop = close.map / total.reads, mismapped.prop = mismapped / total.reads, 
         diff.chr.prop = diff.chr / total.reads, unmapped.prop = unmapped / total.reads)
# Calculate the proportion of each read type for the graph run

long_graph_data = graph_stats %>% 
  select(iteration, divergence, exact.map.prop, close.map.prop, mismapped.prop, diff.chr.prop, unmapped.prop) %>% 
  pivot_longer(cols=c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop"))
# Convert to long format

####################

full_data = rbind(long_bam_data, long_graph_data)
# Combine the BWA and graph data

full_data$name = factor(full_data$name, levels=read_classes)
# Factorize the read classes

cols = corecol(pal="wilke", numcol=6, offset=1)
# Get some colors

overlaps_p = ggplot(full_data, aes(x=iteration, y=value, fill=name)) +
  geom_bar(stat="identity") +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  xlab("Mapping iteration") +
  ylab("Proportion of reads") +
  scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.2)) +
  scale_fill_manual(labels=read_class_names, values=cols) +
  facet_wrap(~divergence) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(overlaps_p)

```


```{r read-lines-graph, out.width="75%", fig.align = "center", warning=FALSE, fig.height=3}

cols = c(corecol(numcol=1, pal="wilke"), corecol(numcol=1, pal="wilke", offset=5), corecol(numcol=3, pal="wilke", offset=1))

exact_p = ggplot(subset(full_data, name == "exact.map.prop"), aes(x=divergence, y=value, color=iteration)) +
  geom_line(size=1) +
  geom_point(size=3, alpha=0.5) +
  xlab("") +
  ylab("Proportion of reads") +
  scale_x_continuous(limits=c(0,0.1), breaks=seq(0, 0.1, by=0.02)) +
  scale_color_manual(name="Iteration", values=cols) +
  bartheme() +
  theme(legend.position="bottom",
        legend.title = element_text(size=12))
#print(read_rates_p)

####################

unmapped_p = ggplot(subset(full_data, name == "unmapped.prop"), aes(x=divergence, y=value, color=iteration)) +
  geom_line(size=1) +
  geom_point(size=3, alpha=0.5) +
  xlab("") +
  ylab("") +
  scale_x_continuous(limits=c(0,0.1), breaks=seq(0, 0.1, by=0.02)) +
  scale_color_manual(name="Iteration", values=cols) +
  bartheme() +
  theme(legend.position="bottom",
        legend.title = element_text(size=12))
#print(read_rates_p)

####################

close_p = ggplot(subset(full_data, name == "close.map.prop"), aes(x=divergence, y=value, color=iteration)) +
  geom_line(size=1) +
  geom_point(size=3, alpha=0.5) +
  xlab("Divergence") +
  ylab("") +
  scale_x_continuous(limits=c(0,0.1), breaks=seq(0, 0.1, by=0.02)) +
  scale_color_manual(name="Iteration", values=cols) +
  bartheme() +
  theme(legend.position="bottom",
        legend.title = element_text(size=12))

####################

leg = get_legend(exact_p)
top = plot_grid(exact_p + theme(legend.position="none"), 
                close_p + theme(legend.position="none"), 
                unmapped_p + theme(legend.position="none"), 
                ncol=3, labels=c("Exact map", "Close map", "Unmapped"))
read_lines_p = plot_grid(top, leg, nrow=2, ncol=1, rel_heights=c(1,0.1))
print(read_lines_p)

```

[< Back ](index.html)

















