---
title: "Mapping simulations"
author: "[gwct](https://gwct.github.io/)"
date: "`r format(Sys.time(), '%m/%d/%Y %H:%M:%S %Z')`"
output:
  #html_document:
  rmdformats::robobook:
    highlight: kate
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
    number_sections: true
    toc_depth: 3
---


```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(ggplot2)
library(cowplot)
library(here)
library(reshape2)
library(dplyr)
library(kableExtra)
source("../lib/design.r")
```

[< Back ](index.html)

# Introduction

Reference genome: mm39

NEAT (divergence) > bcftools consensus > NEAT (heterozygosity) > BWA > GATK

- Chromosomes 18 and 19 only
- chr18 length = 90720763
- chr19 length = 61420004

- Divergence: 0-10%
- Heterozygosity: 0.05%
- Coverage: 30X for both NEAT runs

```{r read, out.width="100%", fig.align = "center", warning=FALSE}
summary_file = here("data", "mm39-30X-vcf-summary.csv")
sum_stats_chr = read.csv(summary_file, header=T)
# Read the summary file from compare_vcfs

sum_stats = sum_stats_chr %>% group_by(divergence) %>% summarize("golden.variants"=sum(golden.variants), "golden.invariant"=sum(golden.invariant), "called.variants"=sum(called.variants), "tp"=sum(tp), "fp"=sum(fp), "fn"=sum(fn), "tn"=sum(tn), "called.het.golden"=sum(called.het.golden), "called.het.non.golden"=sum(called.het.non.golden), "no.info"=sum(no.info))
# Summarize counts by divergence levels (sums counts by chromosome)

sum_stats$het.rate.golden = sum_stats$called.het.golden / sum_stats$golden.variants
sum_stats$het.rate.non.golden = sum_stats$called.het.non.golden / sum_stats$golden.invariant
# Calculate the heterozygosity rate for sites that were simulated to have substitutions and sites that weren't

len_18 = 90720763
len_19 = 61420004
total_len = len_18 + len_19
## TODO: Use .fai file to do this automatically in the full sims

sum_stats$div.rate.golden = sum_stats$golden.variants / total_len
sum_stats$div.rate.called = sum_stats$called.variants / total_len
# Confirm the simulated divergence rates by comparing both the golden and called variants to
# the total length of the genome simulated

sum_stats %>% kable() %>% kable_styling(bootstrap_options=c("striped", "condended", "responsive"), full_width=F)
# Display counts as a table

####################

snp_file = here("data", "mm39-30X-vcf-snps-genmap.csv.gz")
snp_stats_chr = read.csv(gzfile(snp_file), header=T, comment.char="#")
#names(snp_stats_chr)[1] = "coverage"
#snp_stats_chr = subset(snp_stats_chr, coverage != "# coverage")
# Read the snp stats file

####################

mapping_file = here("data", "mm39-30X-bam-summary.csv")
bam_stats = read.csv(mapping_file, header=T)
# Read the bam compare file

####################

```

# Simulated divergence vs. simulated variants

```{r div-v-sim, out.width="50%", fig.align = "center", warning=FALSE}

div_sim_p = ggplot(sum_stats, aes(x=divergence, y=golden.variants)) +
  geom_point(size=3, color=corecol(pal="wilke", numcol=1)) +
  geom_smooth(method="lm", se=F, linetype="dashed", color=corecol(pal="wilke", numcol=1)) +
  #scale_y_continuous(expand=c(0,0)) +
  xlab("Simulated divergence") +
  ylab("# of simulated variants") +
  bartheme()

print(div_sim_p)
```

# Simulated divergence vs. actual % of sites with SNPs simulated

```{r div-v-sim-perc, out.width="50%", fig.align = "center", warning=FALSE}

div_sim_perc_p = ggplot(sum_stats, aes(x=divergence, y=div.rate.golden)) +
  geom_point(size=3, color=corecol(pal="wilke", numcol=1)) +
  geom_smooth(aes(color="Fit"), method="lm", se=F, linetype="dashed") +
  geom_abline(aes(slope=1, intercept=0, color="1:1"), size=1, linetype="dashed", show.legend=F) +
  scale_color_manual(values=c("1:1"="#999999", "Fit"=corecol(pal="wilke", numcol=1))) +
  #scale_y_continuous(expand=c(0,0)) +
  xlab("Simulated divergence") +
  ylab("% of sites with SNPs simulated") +
  bartheme()

print(div_sim_perc_p)
```

# Simulated divergence vs. variants called

After simulating heterozygosity and re-mapping

```{r div-v-called, out.width="50%", fig.align = "center", warning=FALSE}

div_called_p = ggplot(sum_stats, aes(x=divergence, y=called.variants)) +
  geom_point(size=3, color=corecol(pal="wilke", numcol=1)) +
  geom_smooth(method="lm", se=F, linetype="dashed", color=corecol(pal="wilke", numcol=1)) +
  #scale_y_continuous(expand=c(0,0)) +
  xlab("Simulated divergence") +
  ylab("# of called variants") +
  bartheme()

print(div_called_p)
#cat(paste("# Avg. length: ", mean(ratite_data$length)))
#cat(paste("# Median length: ", median(ratite_data$length)))
```

# Simulated divergence vs. % of sites with variants called

After simulating heterozygosity and re-mapping

```{r div-v-called-perc, out.width="50%", fig.align = "center", warning=FALSE}

div_sim_perc_p = ggplot(sum_stats, aes(x=divergence, y=div.rate.called)) +
  geom_point(size=3, color=corecol(pal="wilke", numcol=1)) +
  geom_smooth(aes(color="Fit"), method="lm", se=F, linetype="dashed") +
  geom_abline(aes(slope=1, intercept=0, color="1:1"), size=1, linetype="dashed", show.legend=F) +
  scale_y_continuous(limits=c(0,0.10)) +
  scale_color_manual(values=c("1:1"="#999999", "Fit"=corecol(pal="wilke", numcol=1))) +
  xlab("Simulated divergence") +
  ylab("% of sites with SNPs called") +
  bartheme()

print(div_sim_perc_p)
```

# Simulated variants vs. called variants

```{r sim-v-called, out.width="50%", fig.align = "center", warning=FALSE}

sim_called_p = ggplot(sum_stats, aes(x=golden.variants, y=called.variants)) +
  geom_smooth(aes(color="Fit"), method="lm", se=F, linetype="dashed") +
  geom_abline(aes(slope=1, intercept=0, color="1:1"), size=1, linetype="dashed", show.legend=F) +
  geom_point(size=3, color=corecol(pal="wilke", numcol=1)) +
  scale_color_manual(values=c("1:1"="#999999", "Fit"=corecol(pal="wilke", numcol=1))) +
  #scale_y_continuous(expand=c(0,0)) +
  xlab("# of simulated variants") +
  ylab("# of called variants") +
  bartheme()

print(sim_called_p)
```

# Heterozygosity

Simulated heterozygosity compared to called heterozygous sites at sites with and without substitutions

```{r est-het, out.width="50%", fig.align = "center", warning=FALSE}

cols = c(corecol(numcol=2), "#333333")
names(cols) = c("SNP sites", "Invariant sites", "Simulated")

sim_called_p = ggplot(sum_stats, aes(x=divergence, y=het.rate.non.golden)) +
  #geom_smooth(aes(color="Fit"), method="lm", se=F, linetype="dashed") +
  geom_hline(aes(color="Simulated"), yintercept=0.005, size=1, linetype="dashed", show.legend=F) +
  geom_line(aes(color="SNP sites"), size=1) +
  geom_point(aes(color="SNP sites"), size=3) +
  geom_line(aes(y=het.rate.golden, color="Invariant sites"), size=1) +
  geom_point(aes(y=het.rate.golden, color="Invariant sites"), size=3) +
  scale_color_manual(values=cols) +
  scale_y_continuous(limits=c(0,0.0055)) +
  xlab("Divergence") +
  ylab("% of sites\ncalled heterozygous") +
  bartheme()

print(sim_called_p)
```

# SNP error rates by divergence

SNP calls were classified as follows:

- True positive (TP): Variant in original golden VCF and heterozygous alternate in called VCF
- False positive (FP): Invariant in original golden VCF and heterozygous alternate in called VCF
- False negative (FN): Variant in original golden VCF and heterozygous reference in called VCF
- True negative (TN): Invariant in original golden VCF and heterozygous reference in called VCF

```{r snp-props, out.width="50%", fig.align = "center", warning=FALSE}

sum_stats$total.snps = sum_stats$tp + sum_stats$fp + sum_stats$fn

sum_stats$tp.prop = sum_stats$tp / sum_stats$total.snps
#sum_stats$shared.pos.prop = sum_stats$shared.pos / sum_stats$total.snps
sum_stats$fp.prop = sum_stats$fp / sum_stats$total.snps
sum_stats$fn.prop = sum_stats$fn / sum_stats$total.snps

sum_stats_sum = select(sum_stats, divergence, golden.variants, called.variants, fp.prop, tp.prop, fn.prop)

long_data = melt(sum_stats_sum, id.vars=c("divergence", "golden.variants", "called.variants"))
#long_data = subset(long_data, variable != "no.info")
#long_data = subset(long_data, variable != "fp" & variable != "fn" & variable != "tp")

long_data$text = NA

long_data[long_data$variable=="fp.prop",]$text = sum_stats$fp[sum_stats$divergence==long_data[long_data$variable=="fp.prop",]$divergence]
long_data[long_data$variable=="tp.prop",]$text = sum_stats$tp[sum_stats$divergence==long_data[long_data$variable=="tp.prop",]$divergence]
#long_data[long_data$variable=="shared.pos.prop",]$text = NA
long_data[long_data$variable=="fn.prop",]$text = sum_stats$fn[sum_stats$divergence==long_data[long_data$variable=="fn.prop",]$divergence]

long_data$variable = factor(long_data$variable, levels=c("fp.prop", "tp.prop", "fn.prop"))

cols = c(corecol(pal="wilke", numcol=1, offset=1), "#bda988", corecol(pal="wilke", numcol=1))

overlaps_p = ggplot(long_data, aes(x=divergence, y=value, fill=variable, label=text)) +
  geom_bar(stat="identity") +
  geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  #geom_text(aes(label=text2), size=2.3, color="#f2f2f2", nudge_y=0.05) +
  #geom_text(aes(label=text3), size=2.3, color="#f2f2f2", nudge_y=-0.05) +
  #geom_text(aes(label=text4), size=2.3, color="#f2f2f2", nudge_y=0.82) +
  xlab("Divergence") +
  ylab("Proportion of variants") +
  scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.1)) +
  #scale_x_discrete(limits=unique(long_data$divergence)) +
  scale_x_continuous(breaks=unique(long_data$divergence)) +
  scale_fill_manual(guide=guide_legend(reverse=T), labels=c("False positive", "True positive", "False negative"), values=cols) +
  #scale_fill_discrete(, values=cols) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(overlaps_p)

```



```{r snp-rates, out.width="50%", fig.align = "center", warning=FALSE}

cols = c(corecol(pal="wilke", numcol=1, offset=1), corecol(pal="wilke", numcol=1))

snp_rates_p = ggplot(subset(long_data, variable != "tp.prop" & divergence > 0), aes(x=divergence, y=value, color=variable)) +
  geom_line(size=1.2) +
  geom_point(size=3) +
  xlab("Simulated divergence") +
  ylab("Proportion of variants") +
  scale_x_continuous(breaks=unique(long_data$divergence)) +
  scale_color_manual(guide=guide_legend(reverse=T), labels=c("False positive", "False negative"), values=cols) +
  bartheme() +
  theme(legend.position="bottom")
print(snp_rates_p)

```

# SNP errors and read depth

NAs are called heterozygous sites (either at simulated diverged sites or not)

```{r fp-fn-depth, out.width="50%", fig.align = "center", warning=FALSE}

#snp_stats_chr = subset(snp_stats_chr, !is.na(dp))
#snp_stats_chr$dp = as.numeric(snp_stats_chr$dp)

dp_p = ggplot(snp_stats_chr, aes(x=type, y=dp)) +
  geom_violin(aes(fill=type), alpha=0.3) +
  geom_boxplot(width=0.1, outlier.shape=NA) +
  xlab("SNP classification") +
  ylab("Depth") +
  facet_wrap(~divergence) +
  bartheme() +
  theme(legend.position="none")
print(dp_p)


```

# SNP errors and mapping quality

NAs are called heterozygous sites (either at simulated diverged sites or not)

```{r fp-fn-mq, out.width="50%", fig.align = "center", warning=FALSE}

#snp_stats_chr$mq = as.numeric(snp_stats_chr$mq)

mq_p = ggplot(snp_stats_chr, aes(x=type, y=mq)) +
  geom_violin(aes(fill=type), alpha=0.3) +
  geom_boxplot(width=0.1, outlier.shape=NA) +
  xlab("SNP classification") +
  ylab("Mapping Quality") +
  facet_wrap(~divergence) +
  bartheme() +
  theme(legend.position="none")
print(mq_p)

```

# SNP errors and mappability

Mappability determined with [GenMap's](https://github.com/cpockrandt/genmap) kmer approach with k = 150bp (same as simulated read size).

NAs are called heterozygous sites (either at simulated diverged sites or not)

```{r fp-fn-mb, out.width="50%", fig.align = "center", warning=FALSE}

mb_p = ggplot(snp_stats_chr, aes(x=type, y=mappability)) +
  geom_violin(aes(fill=type), alpha=0.3) +
  geom_boxplot(width=0.1, outlier.shape=NA) +
  xlab("SNP classification") +
  ylab("Mappability") +
  facet_wrap(~divergence) +
  bartheme() +
  theme(legend.position="none")
print(mb_p)

```

# Read mapping errors

Comparing mapped reads to the golden BAM file from the second round of simulation (simulating heterozygosity). Use CrossMap to liftover coordinates from reference genome to simulated genome for mapped reads.

Definitions:

- Exact map = read maps to exact same position as read with same name in golden BAM
- Close map = read maps to within 150bp as read with same name in golden BAM
- Mismapped = read maps to same chromosome as read with same name in golden BAM, but farther than 150bp away
- Different chrome = read maps to different chromosome as read with same name in golden BAM
- Unmapped = read does not map to reference genome
- Missing = read with this name is missing in the mapped BAM file (problem with liftover??)

```{r read-props, out.width="50%", fig.align = "center", warning=FALSE}

bam_stats$total.reads = bam_stats$exact.map + bam_stats$close.map + bam_stats$mismapped + bam_stats$diff.chr + bam_stats$unmapped + bam_stats$missing.in.query

bam_stats$exact.map.prop = bam_stats$exact.map / bam_stats$total.reads
bam_stats$close.map.prop = bam_stats$close.map / bam_stats$total.reads
bam_stats$mismapped.prop = bam_stats$mismapped / bam_stats$total.reads
bam_stats$diff.chr.prop = bam_stats$diff.chr / bam_stats$total.reads
bam_stats$unmapped.prop = bam_stats$unmapped / bam_stats$total.reads
bam_stats$missing.prop = bam_stats$missing.in.query / bam_stats$total.reads

bam_stats_sum = select(bam_stats, divergence, exact.map.prop, close.map.prop, mismapped.prop, diff.chr.prop, unmapped.prop, missing.prop)

long_data = melt(bam_stats_sum, id.vars=c("divergence"))
#long_data = subset(long_data, variable != "coverage")
#long_data = subset(long_data, variable != "exact.map" & variable != "close.map" & variable != "mismapped" & variable != "diff.chr" & variable != "unmapped")

#long_data$text = NA

#long_data$text = bam_stats

#long_data[long_data$variable=="exact.map.prop",]$text = bam_stats$exact.map[bam_stats$divergence==long_data[long_data$variable=="exact.map.prop",]$divergence]
#long_data[long_data$variable=="close.map.prop",]$text = bam_stats$close.map[bam_stats$divergence==long_data[long_data$variable=="close.map.prop",]$divergence]
#long_data[long_data$variable=="mismapped.prop",]$text = bam_stats$mismapped[bam_stats$divergence==long_data[long_data$variable=="mismapped.prop",]$divergence]
#long_data[long_data$variable=="diff.chr.prop",]$text = bam_stats$diff.chr[bam_stats$divergence==long_data[long_data$variable=="diff.chr.prop",]$divergence]
#long_data[long_data$variable=="unmapped.prop",]$text = bam_stats$unmapped[bam_stats$divergence==long_data[long_data$variable=="unmapped.prop",]$divergence]

read_classes = c("exact.map.prop", "close.map.prop", "mismapped.prop", "diff.chr.prop", "unmapped.prop", "missing.prop")
read_class_names = c("Exact map", "Close map", "Mismapped", "Different chrome", "Unmapped", "Missing")

long_data$variable = factor(long_data$variable, levels=rev(read_classes))

cols = corecol(pal="wilke", numcol=length(read_class_names), offset=1)
names(cols) = read_classes

overlaps_p = ggplot(long_data, aes(x=divergence, y=value, fill=variable)) +
  geom_bar(stat="identity") +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  #geom_text(aes(label=text2), size=2.3, color="#f2f2f2", nudge_y=0.05) +
  #geom_text(aes(label=text3), size=2.3, color="#f2f2f2", nudge_y=-0.05) +
  #geom_text(aes(label=text4), size=2.3, color="#f2f2f2", nudge_y=0.82) +
  xlab("Divergence") +
  ylab("Proportion of reads") +
  scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.1)) +
  #scale_x_discrete(limits=unique(long_data$divergence)) +
  scale_x_continuous(breaks=unique(long_data$divergence)) +
  scale_fill_manual(labels=read_class_names, values=cols) +
  #scale_fill_discrete(, values=cols) +
  bartheme() +
  theme(legend.position="bottom") +
  coord_flip()
print(overlaps_p)

```

```{r read-rates, out.width="50%", fig.align = "center", warning=FALSE}

cols = corecol(pal="wilke", numcol=5, offset=2)
names(cols) = read_classes[-1]

read_rates_p = ggplot(subset(long_data, variable != "exact.map.prop"), aes(x=divergence, y=value, color=variable)) +
  geom_line(size=1.2) +
  geom_point(size=3) +
  #geom_text(size=2.3, position=position_stack(vjust=0.5), color="#f2f2f2") +
  #geom_text(aes(label=text2), size=2.3, color="#f2f2f2", nudge_y=0.05) +
  #geom_text(aes(label=text3), size=2.3, color="#f2f2f2", nudge_y=-0.05) +
  #geom_text(aes(label=text4), size=2.3, color="#f2f2f2", nudge_y=0.82) +
  xlab("Divergence") +
  ylab("Proportion of reads") +
  #scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.1)) +
  #scale_x_discrete(limits=unique(long_data$divergence)) +
  scale_x_continuous(breaks=unique(long_data$divergence)) +
  scale_color_manual(labels=read_class_names[-1], values=cols) +
  #scale_fill_discrete(, values=cols) +
  bartheme() +
  theme(legend.position="bottom")
print(read_rates_p)

```

[< Back ](index.html)

















