---
title: "Untitled"
output: html_document
date: "2023-06-13"
---

```{r setup, include=FALSE}

############################################################
# For Mapping sims Evolution talk
# June 2023
# SNP accuracy after 3 iterations
# Gregg Thomas
############################################################

knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(here)
source(here("evolution/scripts/lib/design.r"))

############################################################

total_len = 90720763

############################################################

vcf_comp_file = here("data", "mm39-30X-0.005h-3i-vcf-comparison.tsv.gz")
vcf_comp = read_tsv(vcf_comp_file, col_names=c("coverage", "divergence", "heterozygosity", "iteration", "chr", "pos", "ref", "golden.div", "prev.iter", "iter.gt", "iter.allele.1", "iter.allele.2", "minimap.gt", "minimap.allele.1", "minimap.allele.2"))
# Load the SNP data (~15gb, takes time)

vcf_comp = vcf_comp %>% filter(divergence != 0)

tp = vcf_comp %>% filter(golden.div != ref & minimap.gt == "hom.alt" & minimap.allele.1 == golden.div) %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="tp")
fp = vcf_comp %>% filter(golden.div == ref & minimap.gt == "hom.alt") %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="fp")
fn = vcf_comp %>% filter(golden.div != ref & minimap.gt == "hom.ref" & minimap.allele.1 == ref) %>% 
  group_by(divergence, iteration) %>% 
  summarize(n=n(), class="fn")
# Get the different error classes in separate dfs

############################################################

```


```{r plot, include=FALSE}

snp_classes = rbind(tp, fp, fn)
snp_classes = snp_classes %>% group_by(divergence, iteration) %>% mutate(total.sites = sum(n)) %>% mutate(prop = n / total.sites)
snp_classes$class = factor(snp_classes$class, levels=c("fn", "tp", "fp"))
# Combine and summarize the classes

cols = c("fp"=corecol(pal="wilke", numcol=1, offset=1), "tp"="#bda988", "fn"=corecol(pal="wilke", numcol=1))

snps_p = ggplot(snp_classes, aes(x=iteration, y=prop, fill=class)) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE)) +
  xlab("Mapping iteration") +
  ylab("Proportion of variants") +
  scale_y_continuous(expand=c(0,0), limits=c(0,1), breaks=seq(0, 1, by=0.2)) +
  scale_fill_manual(labels=c("False negative", "True positive", "False positive"), values=cols) +
  facet_wrap(~divergence) +
  bartheme() +
  theme(legend.position="bottom",
        panel.spacing=unit(1.5, "lines"),
        axis.text=element_text(size=12), 
        axis.title=element_text(size=14),
        plot.margin = unit(c(1,0.5,0.25,0.25), "cm"),
        #strip.background=element_blank(),
        strip.text=element_text(size=14)) +
  coord_flip()
print(snps_p)

############################################################

```

```{r save, include=FALSE}
  outfilename = here("evolution", "figs", "05-iterative-snp-accuracy.png")
  ggsave(filename=outfilename, reads_p, width=8, height=5, units="in")
```

