---
title: ''
output: html_document
date: "2023-09-06"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(JBrowseR)
```


```{r}

# ref = "https://rcood.rc.fas.harvard.edu/pun/sys/dashboard/files/fs/n/holylfs05/LABS/informatics/Users/gthomas/Mapping-simulations-data/reference-genomes/mm39/mm39-18.fa"
# bam = "https://rcood.rc.fas.harvard.edu/pun/sys/dashboard/files/fs/n/holylfs05/LABS/informatics/Users/gthomas/Mapping-simulations-data/mm39-18-iterative/mapped-reads/30X/0.10/0.005/iter1/mm39-30X-0.10d-0.005h.bam"

#JBrowseR("View", assembly = ref, tracks = tracks(bam))

#, location="18:3000459..3000469"

#JBrowseR("ViewHg19", location = "10:29,838,737..29,838,819")

ref = assembly("https://rcood.rc.fas.harvard.edu/pun/sys/dashboard/files/fs/n/holylfs05/LABS/informatics/Users/gthomas/Mapping-simulations-data/reference-genomes/mm39/mm39-18.fa")
bam = track_alignments("https://rcood.rc.fas.harvard.edu/pun/sys/dashboard/files/fs/n/holylfs05/LABS/informatics/Users/gthomas/Mapping-simulations-data/mm39-18-iterative/mapped-reads/30X/0.10/0.005/iter1/mm39-30X-0.10d-0.005h.bam", assembly=ref)

#JBrowseR("View", assembly = ref, tracks = tracks(bam), location="18:3000449..3000469")
# 3000459
# Not FN? No reads mapped though?
# grep -m1 3000459 mm39-iter1-18-30X-0.10d-0.005h-compare-vcf-full.tsv
# coverage      divergence      heterozygosity  iteration       chr     pos     ref     golden.div      prev.iter       iter.gt iter.allele.1   iter.allele.2   minimap.gt      minimap.allele.1        minimap.allele.2
# 30      0.10    0.005   1       18      3000459 G       G       G       hom.prev        G       G       hom.ref G       G

#JBrowseR("View", assembly = ref, tracks = tracks(bam), location="18:3006081..3006101")
# 3006091
# Actual FN
# grep -m1 3006091 mm39-iter1-18-30X-0.10d-0.005h-compare-vcf-full.tsv
# coverage      divergence      heterozygosity  iteration       chr     pos     ref     golden.div      prev.iter       iter.gt iter.allele.1   iter.allele.2   minimap.gt      minimap.allele.1        minimap.allele.2
# 30      0.10    0.005   1       18      3006091 G       A       G       hom.prev        G       G       hom.ref G       G

#JBrowseR("View", assembly = ref, tracks = tracks(bam), location="18:3051360..3051380")
# 3051370
# Actual FN
# grep -m1 3051370 mm39-iter1-18-30X-0.10d-0.005h-compare-vcf-full.tsv
# coverage      divergence      heterozygosity  iteration       chr     pos     ref     golden.div      prev.iter       iter.gt iter.allele.1   iter.allele.2   minimap.gt      minimap.allele.1        minimap.allele.2
# 30      0.10    0.005   1       18      3051370 A       G       A       hom.prev        A       A       hom.ref A       A


JBrowseR("View", assembly = ref, tracks = tracks(bam), location="18:90339663..90339683")
# 90339673
# Actual FN
# grep -m1 90339673 mm39-iter1-18-30X-0.10d-0.005h-compare-vcf-full.tsv
# coverage      divergence      heterozygosity  iteration       chr     pos     ref     golden.div      prev.iter       iter.gt iter.allele.1   iter.allele.2   minimap.gt      minimap.allele.1        minimap.allele.2
# 30      0.10    0.005   1       18      90339673        T       C       T       hom.prev        T       T       hom.ref T       T

```



























